kpops/utils/gen_schema.py

```diff
@@ -0,0 +1,139 @@
+import logging
+from enum import Enum
+from typing import (
+    Annotated,
+    Any,
+    Literal,
+    Optional,
+    Sequence,
+    Union,
+    get_args,
+    get_origin,
+)
+
+from pydantic import Field, schema, schema_json_of
+from pydantic.fields import ModelField
+from pydantic.schema import SkipField
+
+from kpops.cli.pipeline_config import PipelineConfig
+from kpops.cli.registry import _find_classes
+from kpops.components.base_components.pipeline_component import PipelineComponent
+
+
+class SchemaScope(str, Enum):
+    PIPELINE = "pipeline"
+    CONFIG = "config"
+
+
+original_field_schema = schema.field_schema
+
+
+# adapted from https://github.com/tiangolo/fastapi/issues/1378#issuecomment-764966955
+def field_schema(field: ModelField, **kwargs: Any) -> Any:
+    if field.field_info.extra.get("hidden_from_schema"):
+        raise SkipField(f"{field.name} field is being hidden")
+    else:
+        return original_field_schema(field, **kwargs)
+
+
+schema.field_schema = field_schema
+
+log = logging.getLogger("")
+
+
+def is_valid_component(
+    used_fields: set[str], component: type[PipelineComponent]
+) -> bool:
+    """
+    Check whether a PipelineComponent subclass has a valid definition for the schema generation.
+
+    :param used_fields: types defined so far
```

Comment:
the parameter name `used_fields` might be a little bit misleading

Reply:
agreed, changed to `defined_component_types`